In [1]:
import pandas as pd
import matplotlib
import plotly.express as px
import plotly.graph_objects as go

/tmp/ipykernel_318355/2714449696.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
df = pd.read_csv('../data/Netflix Userbase.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   User ID            2500 non-null   int64 
 1   Subscription Type  2500 non-null   object
 2   Monthly Revenue    2500 non-null   int64 
 3   Join Date          2500 non-null   object
 4   Last Payment Date  2500 non-null   object
 5   Country            2500 non-null   object
 6   Age                2500 non-null   int64 
 7   Gender             2500 non-null   object
 8   Device             2500 non-null   object
 9   Plan Duration      2500 non-null   object
dtypes: int64(3), object(7)
memory usage: 195.4+ KB


In [4]:
df.head()

User ID Subscription Type  Monthly Revenue Join Date Last Payment Date  \
0        1             Basic               10  15-01-22          10-06-23   
1        2           Premium               15  05-09-21          22-06-23   
2        3          Standard               12  28-02-23          27-06-23   
3        4          Standard               12  10-07-22          26-06-23   
4        5             Basic               10  01-05-23          28-06-23   

          Country  Age  Gender      Device Plan Duration  
0   United States   28    Male  Smartphone       1 Month  
1          Canada   35  Female      Tablet       1 Month  
2  United Kingdom   42    Male    Smart TV       1 Month  
3       Australia   51  Female      Laptop       1 Month  
4         Germany   33    Male  Smartphone       1 Month

## Verificacion Cantidad Null

In [5]:
print(df.shape)
df.isnull().sum()

(2500, 10)


User ID              0
Subscription Type    0
Monthly Revenue      0
Join Date            0
Last Payment Date    0
Country              0
Age                  0
Gender               0
Device               0
Plan Duration        0
dtype: int64

In [6]:
def nulos(df):
    porcentajes_nulos = (df.isnull().mean() * 100).round(2)
    resumen_nulos = pd.DataFrame({'Columna': porcentajes_nulos.index, 'Porcentaje Nulos': porcentajes_nulos.values})
    resumen_nulos = resumen_nulos.sort_values(by='Porcentaje Nulos', ascending=False)

    print(resumen_nulos)

nulos(df)

             Columna  Porcentaje Nulos
0            User ID               0.0
1  Subscription Type               0.0
2    Monthly Revenue               0.0
3          Join Date               0.0
4  Last Payment Date               0.0
5            Country               0.0
6                Age               0.0
7             Gender               0.0
8             Device               0.0
9      Plan Duration               0.0


## Analisis descriptivo

In [7]:
df.describe()

User ID  Monthly Revenue          Age
count  2500.00000      2500.000000  2500.000000
mean   1250.50000        12.508400    38.795600
std     721.83216         1.686851     7.171778
min       1.00000        10.000000    26.000000
25%     625.75000        11.000000    32.000000
50%    1250.50000        12.000000    39.000000
75%    1875.25000        14.000000    45.000000
max    2500.00000        15.000000    51.000000

User ID	
- Count: Hay 2500 registros en total.
- Desviación estándar: La desviación estándar de los identificadores de usuario es de aproximadamente 721.83.
- Mínimo: El valor mínimo de los identificadores de usuario es 1.
- Percentiles: El 25% de los identificadores de usuario están por debajo de 625.75, el 50% están por debajo de 1250.5 y el 75% están por debajo de 1875.25.
- Máximo: El valor máximo de los identificadores de usuario es 2500.

Monthly Revenue	
- Media: El ingreso mensual promedio es de aproximadamente 12.51.
- Desviación estándar: La desviación estándar del ingreso mensual es de aproximadamente 1.69.
- Mínimo: El ingreso mensual mínimo es de 10.
- Percentiles: El 25% de los ingresos mensuales están por debajo de 11, el 50% están por debajo de 12 y el 75% están por debajo de 14.
- Máximo: El ingreso mensual máximo es de 15.

Age
- Media: La edad promedio es de aproximadamente 38 años.
- Desviación estándar: La desviación estándar de las edades es de aproximadamente 7.
- Mínimo: La edad mínima es de 26 años.
- Percentiles: El 25% de las edades están por debajo de 32, el 50% están por debajo de 39 y el 75% están por debajo de 45.
- Máximo: La edad máxima es de 51 años.

## Representacion valores NaN

In [8]:
is_null_df = pd.isnull(df)
fig = px.imshow(is_null_df,
                labels=dict(color='NaN'),
                color_continuous_scale=['white', 'darkred'],  
                title='Mapa de Calor de Valores NaN ')

fig.update_layout(coloraxis_colorbar=dict(tickvals=[0.25, 0.75], ticktext=['No NaN', 'NaN']),
                  coloraxis=dict(cmin=-0.5, cmax=1))

## Promedio generos y calcula el ingreso mensual promedio por paises

In [9]:
grouped_data = df.groupby(['Country', 'Gender']).agg({'Monthly Revenue': 'mean'})
grouped_data

Monthly Revenue
Country        Gender                 
Australia      Female        12.370787
               Male          12.446809
Brazil         Female        12.168421
               Male          12.829545
Canada         Female        12.566879
               Male          12.356250
France         Female        12.593407
               Male          12.619565
Germany        Female        12.212766
               Male          12.494382
Italy          Female        12.615385
               Male          12.706522
Mexico         Female        12.258427
               Male          12.191489
Spain          Female        12.562232
               Male          12.545872
United Kingdom Female        12.752688
               Male          12.577778
United States  Female        12.711111
               Male          12.407080

In [10]:
# Datos
countries = ['Australia', 'Brazil', 'Canada', 'France', 'Germany', 'Italy', 'Mexico', 'Spain', 'United Kingdom', 'United States']
genders = ['Female', 'Male']
monthly_revenue = [
    [12.370787, 12.446809],
    [12.168421, 12.829545],
    [12.566879, 12.356250],
    [12.593407, 12.619565],
    [12.212766, 12.494382],
    [12.615385, 12.706522],
    [12.258427, 12.191489],
    [12.562232, 12.545872],
    [12.752688, 12.577778],
    [12.711111, 12.407080]
]

# Crear el gráfico
fig = go.Figure()

for i, country in enumerate(countries):
    for j, gender in enumerate(genders):
        fig.add_trace(go.Bar(
            x=[country],
            y=[monthly_revenue[i][j]],
            name=gender,
            marker_color='blue' if gender == 'Female' else 'orange'
        ))

# Ajustes del diseño
fig.update_layout(
    title='Ingresos Mensuales Promedio por País y Género',
    xaxis=dict(title='País'),
    yaxis=dict(title='Ingreso Mensual Promedio'),
    barmode='group',
    legend=dict(title='Género')
)

# Mostrar el gráfico
fig.show()

### Convertir columna fecha a su tipo de dato para graficacion de inscripciones por fechas

In [ ]:
df['Join Date'] = pd.to_datetime(df['Join Date'], format='%d-%m-%y')
df['Last Payment Date'] = pd.to_datetime(df['Last Payment Date'], format='%d-%m-%y')

### Inscripciones por mes

In [16]:
def asignar_semestre(fecha):
    if fecha.month <= 6:
        return '1er Semestre'
    else:
        return '2do Semestre'

df['Semestre de Join'] = df['Join Date'].apply(asignar_semestre)

inscripciones_por_semestre = df.groupby('Semestre de Join').size().reset_index(name='Inscripciones')

fig = px.bar(inscripciones_por_semestre, x='Semestre de Join', y='Inscripciones', title='Inscripciones de usuarios por semestre de inscripción')
fig.show()

### Tipos de suscripciones

In [17]:
fig = px.bar(df, x='Subscription Type', title='Distribución de Subscription Type')
fig.show()

### Histograma por rangos de edad

In [18]:
fig_hist_age = px.histogram(df, x='Age', title='Histograma de Age')
fig_hist_age.show()

### Asignacion de coordenadas por region de cada Pais

In [24]:
coordinates = {
    'Country': ['Australia', 'Brazil', 'Canada', 'France', 'Germany', 'Italy', 'Mexico', 'Spain', 'United Kingdom', 'United States'],
    'Latitude': [-25.2744, -14.2350, 56.1304, 46.6034, 51.1657, 41.8719, 23.6345, 40.4637, 55.3781, 37.7749],
    'Longitude': [133.7751, -51.9253, -106.3468, 1.8883, 10.4515, 12.5674, -102.5528, -3.7492, -3.4360, -122.4194]
}
coordinates_df = pd.DataFrame(coordinates)

df_with_coordinates = pd.merge(df, coordinates_df, on='Country')

fig_map = px.density_mapbox(df_with_coordinates, lat='Latitude', lon='Longitude', z='Monthly Revenue', radius=10,
                            center=dict(lat=0, lon=180), zoom=0, mapbox_style="carto-positron",
                            title='Distribución de Monthly Revenue por país')
fig_map.show()

In [23]:
df['Join Date'] = pd.to_datetime(df['Join Date'])
df['Year'] = df['Join Date'].dt.year
fig_line = px.line(df.groupby('Year').size().reset_index(), x='Year', y=0, title='Inscripciones de usuarios por año')
fig_line.show()

In [25]:
df['Year'] = df['Join Date'].dt.year

df_inscriptions_per_year = df.groupby('Year').size().reset_index(name='Inscriptions')
df_inscriptions_per_year

Year  Inscriptions
0  2021            14
1  2022          2448
2  2023            38